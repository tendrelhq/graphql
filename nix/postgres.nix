{inputs, ...}: {
  imports = [
    inputs.process-compose.flakeModule
  ];
  perSystem = {
    config,
    self',
    pkgs,
    lib,
    ...
  }: {
    devShells.postgres = pkgs.mkShellNoCC {
      name = "tendrelhq/graphql+postgres";
      buildInputs = [
        config.packages.postgresql
        pkgs.perlPackages.TAPParserSourceHandlerpgTAP # pg_prove
        pkgs.python3.pkgs.sqlfmt
        pkgs.sqitchPg
        pkgs.squawk
      ];
    };

    packages = {
      inherit (pkgs) postgresql_14; # beta is on 14.12
      postgresql = pkgs.postgresql_16;
      pgddl = pkgs.stdenv.mkDerivation rec {
        name = "pgddl";
        version = "0.29";
        src = pkgs.fetchFromGitHub {
          owner = "lacanoid";
          repo = name;
          rev = version;
          hash = "sha256-W3G6TGtkj+zXXdGZZR0bmZhsLuFJvuGTlDoo8kL8sf0=";
        };
        buildInputs = with pkgs; [perl config.packages.postgresql];
        postPatch = ''
          patchShebangs .
        '';
        installPhase = ''
          install -D -t $out/share/postgresql/extension *.sql
          install -D -t $out/share/postgresql/extension *.control
        '';
      };
    };

    process-compose.devenv = {
      imports = [
        inputs.services.processComposeModules.default
      ];

      cli.options.no-server = false;

      services.postgres.pg1 = {
        enable = true;
        package = config.packages.postgresql;
        #
        extensions = exts:
          with exts;
          with config.packages; [
            pg_cron
            pgddl
            pgtap
            plpgsql_check
          ];
        hbaConf = lib.mkForce [
          # This one is the default docker bridge network.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.17.0.0/16";
            method = "trust";
          }
          # On my machine this is the tendrel_default network. YMMV.
          {
            type = "host";
            database = "all";
            user = "all";
            address = "172.18.0.0/16";
            method = "trust";
          }
        ];
        initialDatabases = [
          {
            name = "dev";
          }
        ];
        settings = {
          listen_addresses = "0.0.0.0"; # so docker0 can connect
          log_statement = "all";
          logging_collector = false;
          shared_preload_libraries = "pg_cron,plpgsql_check";
        };
      };

      services.pgadmin.pga1 = {
        enable = true;
        extraConfig = {
          # Run in desktop mode with the master password disabled.
          MASTER_PASSWORD_REQUIRED = false;
          SERVER_MODE = false;
        };
        initialEmail = "";
        initialPassword = "";
      };

      settings.processes = {
        pga1 = {
          depends_on.pg1.condition = "process_healthy";
        };
      };
    };
  };
}
