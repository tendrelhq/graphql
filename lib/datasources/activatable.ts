import type { ID, ResolversTypes } from "@/schema";
import { decodeGlobalId } from "@/schema/system";
import type { WithKey } from "@/util";
import DataLoader from "dataloader";
import type { Request } from "express";
import { match } from "ts-pattern";
import { sql, unionAll } from "./postgres";

export function makeActivatableLoader(_req: Request) {
  return new DataLoader<ID, ResolversTypes["Activatable"] | undefined>(
    async keys => {
      const entities = keys.map(decodeGlobalId);
      const byUnderlyingType = entities.reduce((acc, { type, id }) => {
        if (!acc.has(type)) acc.set(type, []);
        acc.get(type)?.push(id);
        return acc;
      }, new Map<string, string[]>());

      const qs = [...byUnderlyingType.entries()].flatMap(([type, ids]) =>
        match(type)
          .with(
            "workresult",
            () => sql`
                (
                  WITH cte AS (
                      SELECT
                          wr.id AS _key,
                          encode(('workresult:' || wr.id || ':active')::bytea, 'base64') AS id,
                          (
                              wr.workresultenddate IS NULL
                              OR wr.workresultenddate > now()
                          ) AS active,
                          wr.workresultstartdate,
                          wr.workresultenddate
                      FROM public.workresult AS wr
                      WHERE wr.id IN ${sql(ids)}
                  )

                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              'Instant' AS "__typename",
                              (extract(epoch from workresultstartdate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      )
                  FROM cte
                  WHERE active = true
                  UNION ALL
                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              'Instant' AS "__typename",
                              (extract(epoch from workresultenddate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      )
                  FROM cte
                  WHERE active = false
                )
            `,
          )
          .with(
            "worktemplate",
            () => sql`
                (
                  WITH cte AS (
                      SELECT
                          id AS _key,
                          encode(('worktemplate:' || id || ':active')::bytea, 'base64') AS id,
                          (
                              worktemplateenddate IS NULL
                              OR worktemplateenddate > now()
                          ) AS active,
                          worktemplatestartdate,
                          worktemplateenddate
                      FROM public.worktemplate
                      WHERE id IN ${sql(ids)}
                  )

                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              SELECT
                                  'Instant' AS "__typename",
                                  (extract(epoch from worktemplatestartdate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      ) AS "updatedAt"
                  FROM cte
                  WHERE active = true
                  UNION ALL
                  SELECT
                      _key,
                      id,
                      active,
                      (
                          SELECT row_to_json(t)
                          FROM (
                              SELECT
                                  'Instant' AS "__typename",
                                  (extract(epoch from worktemplateenddate) * 1000)::text AS "epochMilliseconds"
                          ) t
                      ) as "updatedAt"
                  FROM cte
                  WHERE active = false
                )
            `,
          )
          .otherwise(() => []),
      );

      if (!qs.length) return entities.map(() => undefined);

      type X = WithKey<ResolversTypes["Activatable"]>;
      const xs = await sql<X[]>`${unionAll(qs)}`;
      return entities.map(e => xs.find(x => e.id === x._key));
    },
  );
}
