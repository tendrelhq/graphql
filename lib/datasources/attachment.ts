import type { Attachment } from "@/schema";
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import Dataloader from "dataloader";
import type { Request } from "express";
import { sql } from "./postgres";
import { decodeGlobalId } from "@/schema/system";
import { GraphQLError } from "graphql";

async function createPresignedUrl(client: S3Client, uri: URL) {
  const command = new GetObjectCommand({
    Bucket: uri.host,
    Key: uri.pathname.slice(1), // remove the leading '/'
  });
  if (uri.protocol !== "s3:") {
    throw "invariant violated";
  }
  return getSignedUrl(client, command, {
    expiresIn: Number(process.env.ATTACHMENT_EXPIRATION_TIME_SECONDS ?? 3600),
  });
}

export default (_: Request) => ({
  byId: new Dataloader<string, Attachment>(async keys => {
    const pks = keys.map(k => decodeGlobalId(k).id);
    const rows = await sql<{ _key: string; uri: string }[]>`
        SELECT
            workpictureinstanceuuid AS _key,
            workpictureinstancestoragelocation AS uri
        FROM public.workpictureinstance
        WHERE workpictureinstanceuuid IN ${sql(pks)};
    `;
    const s3 = new S3Client();
    return Promise.all(
      pks.map(async (pk, i) => {
        const row = rows.find(r => r._key === pk);

        if (!row) {
          throw new GraphQLError(`No Attachment for id '${pk}'`, {
            extensions: {
              code: "NOT_FOUND",
            },
          });
        }

        return {
          id: keys[i],
          attachment: await createPresignedUrl(s3, new URL(row.uri)),
        };
      }),
    );
  }),
});
