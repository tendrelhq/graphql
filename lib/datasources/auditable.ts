import type { ID, ResolversTypes } from "@/schema";
import { decodeGlobalId, type GlobalId } from "@/schema/system";
import type { WithKey } from "@/util";
import DataLoader from "dataloader";
import type { Request } from "express";
import { match } from "ts-pattern";
import { sql, unionAll } from "./postgres";

export function makeAuditableLoader(_req: Request) {
  return new DataLoader<ID, ResolversTypes["Auditable"] | undefined>(
    async keys => {
      const entities = keys.map(decodeGlobalId);
      const byUnderlyingType = entities.reduce((acc, { type, ...ids }) => {
        if (!acc.has(type)) acc.set(type, []);
        acc.get(type)?.push(ids);
        return acc;
      }, new Map<string, Omit<GlobalId, "type">[]>());

      const qs = [...byUnderlyingType.entries()].flatMap(([type, ids]) =>
        match(type)
          .with(
            "workresult",
            () => sql`
                SELECT
                    id AS _key,
                    'Auditable' AS "__typename",
                    encode(('workresult:' || id || ':auditable')::bytea, 'base64') AS id,
                    workresultforaudit AS auditable
                FROM public.workresult
                WHERE id IN ${sql(ids.map(i => i.id))}
            `,
          )
          .with(
            "worktemplate",
            () => sql`
                SELECT
                    id AS _key,
                    'Auditable' AS "__typename",
                    encode(('worktemplate:' || id || ':auditable')::bytea, 'base64') AS id,
                    worktemplateisauditable AS auditable
                FROM public.worktemplate
                WHERE id IN ${sql(ids.map(i => i.id))}
            `,
          )
          .otherwise(() => []),
      );

      if (!qs.length) return entities.map(() => undefined);

      type X = WithKey<ResolversTypes["Auditable"]>;
      const xs = await sql<X[]>`${unionAll(qs)}`;
      return entities.map(e => xs.find(x => e.id === x._key));
    },
  );
}
