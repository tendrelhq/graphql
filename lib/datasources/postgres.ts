import type { Request } from "express";
import z from "myzod";
import postgres, { type Fragment } from "postgres";
import { makeActiveLoader } from "./activatable";
import makeAttachmentLoader from "./attachment";
import { makeAuditableLoader } from "./auditable";
import makeCustomerRequestedLanguageLoader from "./crl";
import { makeDescriptionLoader } from "./description";
import { makeDynamicStringLoader } from "./dynamicString";
import makeInvitationLoader from "./invitation";
import makeLanguageLoader from "./language";
import makeLocationLoader from "./location";
import {
  makeDisplayNameLoader,
  makeNameLoader,
  makeNameMetadataLoader,
} from "./name";
import makeOrganizationLoader from "./organization";
import { makeRequirementLoader } from "./requirement";
import { makeSopLoader } from "./sop";
import { makeStatusLoader } from "./status";
import makeTagLoader from "./tag";
import makeUserLoader from "./user";
import makeWorkerLoader from "./worker";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT: z.number({ coerce: true }).default(10),
  })
  .parse(process.env, { allowUnknown: true });

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  // TODO: this is probably a good idea...
  // connection: {
  //   statement_timeout: DB_STATEMENT_TIMEOUT * 1000, // milliseconds
  // },
});

export type SQL = typeof sql;
export type TxSQL = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}

export function orm(req: Request) {
  return {
    active: makeActiveLoader(req),
    attachment: makeAttachmentLoader(req),
    auditable: makeAuditableLoader(req),
    crl: makeCustomerRequestedLanguageLoader(req),
    description: makeDescriptionLoader(req),
    displayName: makeDisplayNameLoader(req),
    dynamicString: makeDynamicStringLoader(req),
    invitation: makeInvitationLoader(req),
    language: makeLanguageLoader(req),
    location: makeLocationLoader(req),
    name: makeNameLoader(req),
    nameMetadata: makeNameMetadataLoader(req),
    organization: makeOrganizationLoader(req),
    requirement: makeRequirementLoader(req),
    sop: makeSopLoader(req),
    status: makeStatusLoader(req),
    tag: makeTagLoader(req),
    user: makeUserLoader(req),
    worker: makeWorkerLoader(req),
  };
}

export type ORM = ReturnType<typeof orm>;
