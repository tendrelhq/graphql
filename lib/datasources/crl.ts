import { EntityNotFound } from "@/errors";
import type { EnabledLanguage } from "@/schema";
import type { WithKey } from "@/util";
import Dataloader from "dataloader";
import type { Request } from "express";
import { sql } from "./postgres";

export default (_: Request) =>
  new Dataloader<string, EnabledLanguage>(async keys => {
    const rows = await sql<WithKey<EnabledLanguage>[]>`
        SELECT
            customerrequestedlanguageuuid AS _key,
            encode(('enabled-language:' || customerrequestedlanguageuuid)::bytea, 'base64') AS id,
            systaguuid AS "languageId",
            (customerrequestedlanguagelanguageid = customerlanguagetypeid) AS "primary"
        FROM public.customerrequestedlanguage
        INNER JOIN public.customer
            ON customerrequestedlanguagecustomerid = customerid
        INNER JOIN public.systag
            ON customerrequestedlanguagelanguageid = systagid
        WHERE customerrequestedlanguageuuid IN ${sql(keys)};
    `;

    const byKey = rows.reduce(
      (acc, row) => acc.set(row._key as string, row),
      new Map<string, EnabledLanguage>(),
    );

    return keys.map(
      key => byKey.get(key) ?? new EntityNotFound("enabled-language"),
    );
  });
