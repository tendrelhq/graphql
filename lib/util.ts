import type { SortOrder } from "@/schema";
import { type GlobalId, decodeGlobalId } from "@/schema/system";
import { GraphQLError } from "graphql";
import type { Fragment } from "postgres";
import { sql } from "./datasources/postgres";

export function isError<T>(e: T | Error): e is Error {
  return e instanceof Error;
}

export function isValue<T>(v: T | Error): v is T {
  return !isError(v);
}

export function nullish<T>(t: T | null | undefined): t is null | undefined {
  return typeof t === "undefined" || t === null;
}

export class AssertionError extends Error {}

/**
 * Raises an assertion error when the condition does not hold, except in
 * production (and when DISABLE_ASSERTIONS is set) in which case we just log a
 * message instead of throwing.
 */
export function assert(condition: boolean, message = "assertion failed") {
  if (!condition) {
    if (
      // Only fire assertions when in dev/test environments,
      // and only when DISABLE_ASSERTIONS is unset.
      process.env.NODE_ENV !== "production" &&
      typeof process.env.DISABLE_ASSERTIONS === "undefined"
    ) {
      throw new AssertionError(message);
    }

    // Else just log a warning. We don't want to wholly lose our assertions.
    console.debug(`invariant violated: ${message}`);
  }
}

export function assertNonNull<T>(
  value: T | null | undefined,
  message = "Cannot return null for semantically non-nullable field.",
): T {
  if (nullish(value)) {
    throw new Error(message);
  }
  return value;
}

export function assertUnderlyingType(
  expected: string | string[],
  received: string,
) {
  const valid =
    typeof expected === "string"
      ? expected === received
      : expected.some(e => e === received);
  if (!valid) {
    throw new Error(
      `Invalid typename; expected: ${expected}, received: ${received}`,
    );
  }
  return received;
}

export type WithKey<T> = T & { _key: string };

export type PaginationArgs = {
  cursor?: GlobalId | null;
  direction: "forward" | "backward";
  limit: number;
};

type RawPaginationArgs = {
  // forward
  first?: number | null;
  after?: string | null;
  // backward
  last?: number | null;
  before?: string | null;
};

export function buildPaginationArgs(
  args: RawPaginationArgs,
  opts: {
    defaultLimit: number;
    maxLimit: number;
  },
): PaginationArgs {
  validatePaginationArgs(args);

  if (args.first) {
    return {
      cursor: map(args.after, decodeGlobalId),
      direction: "forward",
      limit: Math.min(args.first, opts.maxLimit),
    };
  }

  if (args.last) {
    return {
      cursor: map(args.before, decodeGlobalId),
      direction: "backward",
      limit: Math.min(args.last, opts.maxLimit),
    };
  }

  return {
    direction: "forward",
    limit: opts.defaultLimit,
  };
}

export function sortOrder(s: SortOrder): Fragment {
  switch (s) {
    case "asc":
      return sql`ASC`;
    case "desc":
      return sql`DESC`;
  }
}

export function validatePaginationArgs(args: RawPaginationArgs) {
  if ((args.after && args.before) || (args.first && args.last)) {
    throw new GraphQLError(
      "Invalid pagination arguments. To paginate forward, use the `first` and `after` arguments. To paginate backward, use the `last` and `before` arguments.",
      {
        extensions: {
          code: "BAD_REQUEST",
        },
      },
    );
  }
}

export function map<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
): T extends null | undefined ? T : R {
  if (nullish(t)) {
    return t as T extends null | undefined ? T : R;
  }
  return fn(t as NonNullable<T>) as T extends null | undefined ? T : R;
}

type OrElse<R> = () => R;
export function mapOrElse<T, R>(
  t: T,
  fn: (t: NonNullable<T>) => R,
  orElse: R | OrElse<R>,
): R {
  if (nullish(t)) {
    return typeof orElse === "function" ? (orElse as OrElse<R>)() : orElse;
  }
  return fn(t as NonNullable<T>);
}

export function inspect<T>(t: T) {
  console.debug("t =:", t);
  return t;
}

/**
 * Use this function to compare base64 encoded strings for identity.
 */
export function compareBase64(a: string, b: string): boolean {
  return normalizeBase64(a) === normalizeBase64(b);
}

// Postgres inserts newlines after 76 characters. Apparently it's a thing...
const newlinePattern = /\n/g;

/**
 * Use this function to normalize a base64 encoded string, i.e. remove newline
 * characters that are inserted by Postgres (for example) in certain cases.
 *
 * Note that most base64 implementations simply ignore these seemingly erroneous
 * newlines. As such, this function is mainly useful for (printf) debugging.
 */
export function normalizeBase64(s: string) {
  return s.replace(newlinePattern, "");
}
