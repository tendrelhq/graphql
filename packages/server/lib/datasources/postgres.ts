import postgres, { type Fragment } from "postgres";
import z from "zod";

const { DB_MAX_CONNECTIONS, DB_STATEMENT_TIMEOUT_MS } = z
  .object({
    DB_MAX_CONNECTIONS: z.number({ coerce: true }).default(3),
    DB_STATEMENT_TIMEOUT_MS: z.number({ coerce: true }).default(10_000),
  })
  .parse(process.env);

export const sql = postgres({
  max: DB_MAX_CONNECTIONS,
  connection: {
    // We do some recursive queries and are, in general, not particularly
    // optimized so this is just a last resort to avoid bricking the backend.
    statement_timeout: DB_STATEMENT_TIMEOUT_MS,
  },
  types: {
    bigint: postgres.BigInt,
  },
});

export type Sql = typeof sql;
export type TxSql = Parameters<Parameters<typeof sql.begin>[1]>[0];

/**
 * Like Array.prototype.join but for sql.Fragments.
 */
export function join(xs: readonly Fragment[], d: Fragment) {
  return xs.reduce((acc, x, i) => sql`${acc} ${i ? sql`${d} ${x}` : x}`, sql``);
}

/**
 * Type predicate for whether to include a user input in a dynamic update
 * clause. Note that we only check for `undefined`, and not `null`, because we
 * say that `null` indicates we want to set the underlying database column to
 * NULL.
 */
export function shouldUpdate<T>(input?: T, existing?: T): input is T {
  return typeof input !== "undefined" && input !== existing;
}

export function unionAll(xs: readonly Fragment[]) {
  return join(xs, sql`UNION ALL`);
}
