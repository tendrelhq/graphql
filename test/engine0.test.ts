import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { sql } from "@/datasources/postgres";
import { assert } from "@/util";

describe("engine0", () => {
  let CUSTOMER: string;
  let INSTANCE: string;

  test("build", async () => {
    const result = await sql`
      select target, target_type, row_to_json(t.*) as condition
      from
          engine0.plan_build(task_id := ${INSTANCE}) as p,
          unnest(p.ops) as t
    `;
    expect(result).toBeArrayOfSize(2);
  });

  test("build + check", async () => {
    // set the instance to 'in progress'
    await sql`
      update public.workinstance
      set workinstancestatusid = 707
      where id = ${INSTANCE}
    `;

    const result = await sql`
      select pc.*
      from
          engine0.plan_build(${INSTANCE}) as pb,
          engine0.plan_check(
              target := pb.target,
              target_type := pb.target_type,
              conditions := pb.ops
          ) as pc
    `;
    expect(result).toMatchSnapshot();
  });

  test("build + check + execute", async () => {
    await sql.begin(async tx => {
      const result = await tx`select * from engine0.execute(${INSTANCE})`;
      expect(result.at(0)?.instance).toBeTruthy();
      await tx`rollback and chain`; // to avoid 25P01 error
    });
  });

  beforeAll(async () => {
    const logs = await sql`
      select *
      from mft.create_demo(
          customer_name := 'Frozen Tendy Factory',
          admins := (
              select array_agg(workeruuid)
              from public.worker
              where workerfullname = 'Jerry Garcia'
          )
      )
    `;

    CUSTOMER = logs.find(({ op }) => op.trim() === "+customer")?.id;
    assert(!!CUSTOMER);

    INSTANCE = logs.find(({ op }) => op.trim() === "+instance")?.id;
    assert(!!INSTANCE);

    console.debug("setup:", { CUSTOMER, INSTANCE });
  });

  afterAll(async () => {
    assert(!!CUSTOMER);
    process.stderr.write("Cleaning up... ");
    const [row] = await sql`
      select mft.destroy_demo(${CUSTOMER}) as ok;
    `;
    console.debug(row.ok);
  });
});

// test("fc tutorial", () => {
//   fc.assert(
//     fc.property(fc.array(fc.integer()), data => {
//       const sorted = data.toSorted((a, b) => a - b);
//       for (let i = 1; i < data.length; ++i) {
//         expect(sorted[i - 1]).toBeLessThanOrEqual(sorted[i]);
//       }
//     }),
//     { numRuns: 100_000 },
//   );
// });
