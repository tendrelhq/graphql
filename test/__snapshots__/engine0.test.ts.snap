// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`engine0 build + check 1`] = `
[
  {
    "result": true,
    "system": "engine0.eval_state_condition",
  },
  {
    "result": true,
    "system": "engine0.eval_state_condition",
  },
]
`;
